### DISCUSSION TITLE
are multiple START or COMPLETE events allowed?

### DISCUSSION THREAD

**Discussion by cerveada** [OpenLineage/OpenLineage#597] [Q&A]
As I understand the documentation per one runId there suppose to be: 
- **exactly one** `START` event 
- after that **exactly one** of `[ COMPLETE, ABORT, FAIL ]` event 
- and after that only OTHER event are allowed in any quantity.

But, the `openlineage-spark` is sending multiple START and COMPLETE events. So is it a bug or have I misunderstood the specification?

**Comment by mobuchowski** [comment]
https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.md specifies that 

>Run Event: an event describing an observed state of a job run. It is required to at least send a START event and a COMPLETE/FAIL/ABORT event. Additional events are optional.

OTHER event is not required to be send in any particular time - but, for example, sending it before start would be weird :)
We, for example, want to use them to provide updates for ongoing run in streaming pipelines.

**Reply by cerveada** [reply]
Ok, I understand that there must be at least one START, but that still doesn't answer whether it is valid to send multiple START events.

We want to implement an open lineage API for our application. Without knowing what should be considered error and what is valid, we would have to guess, and that can makes us incompatible with other implementations.

Unfortunately, the documentation seems to be a bit vague about the event lifecycle. It would help to know how many events of each type I must be ready to process and in what order.

**Reply by mobuchowski** [reply]
>Unfortunately, the documentation seems to be a bit vague about the event lifecycle.

That's very valid observation. I've created an issue to clarify it: https://github.com/OpenLineage/OpenLineage/issues/599
Please, add questions in comment if the 4 points I created don't cover everything what we're talking here, and we'll try to make it more precise and obvious.

Anyway, biggest assumption OpenLineage makes for events corresponding to same runId is that they are cumulative. If some information exists in preceding event, let's say START, and isn't present in COMPLETE, then the START information is still valid. 
If the same information exists in both events, then later event (by event time) has the correct information that should "overwrite" earlier one.

The multiple START/COMPLETE events in Spark integration takes advantage of that assumption. We're listening to events in different phase of execution, and combine information available on different execution stages.

**Comment by collado-mike** [comment]
I've added https://github.com/OpenLineage/OpenLineage/issues/636 which addresses this
